From d0789abd663eb9b3b9a3efe0c5f7ad84e6187198 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 18 Jun 2021 09:10:46 +0300
Subject: [PATCH] [SW-33447] greco: handle flr interrupt

The handling of FLR interrupt is different for greco compared to
gaudi. This patch implements greco specific FLR handling function
and implements FLR handling sequence recommended by HW team.

Change-Id: I381a93b53902d2d880ed9d059fb4b054e0e837cd
---

diff --git a/arch/greco/common.c b/arch/greco/common.c
index 9af620a..f31bf44 100644
--- a/arch/greco/common.c
+++ b/arch/greco/common.c
@@ -19,6 +19,8 @@
 #include <pcie_phy_masks.h>
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
 
 #include <cpu_timestamp_regs.h>
 #include <psoc_arc0_cfg_regs.h>
@@ -696,6 +698,20 @@
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
+void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+
+	/*
+	 * Clear PCIe SPI INTR. This is required to clear
+	 * FLR interrupt as recommended by HW team.
+	 */
+	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
+			mmCPU_IF_PCIE_SPI_INTR_CLR);
+}
+
 bool preboot_is_valid_address_range(u64 addr, int len)
 {
 	if (addr >= HL_SP_SRAM_BASE &&
diff --git a/arch/greco_arc_gnu/common.c b/arch/greco_arc_gnu/common.c
index 5d73a22..9016f10 100644
--- a/arch/greco_arc_gnu/common.c
+++ b/arch/greco_arc_gnu/common.c
@@ -21,6 +21,8 @@
 #include <pcie_dbi_masks.h>
 #include <cpu_ca53_cfg_regs.h>
 #include <cpu_timestamp_regs.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
 
 #include <cpu_timestamp_regs.h>
 
@@ -229,6 +231,20 @@
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
+void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+
+	/*
+	 * Clear PCIe SPI INTR. This is required to clear
+	 * FLR interrupt as recommended by HW team.
+	 */
+	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
+			mmCPU_IF_PCIE_SPI_INTR_CLR);
+}
+
 static struct hl_ts_cfg greco_hl_ts_data[] = {
 	{.base = mmDCORE0_MME_TSTD_BASE, .alarm_a = 102, .alarm_b = 120,
 		.use_calib = false, .calib_offset = 10, .hl_tmp_data = { 0 }},
